src/perf/lib/PerfServer.cpp

```diff
@@ -423,3 +569,141 @@ PerfServer::TcpSendCompleteCallback(
+    // Not using the high priority flag for delay threads
+    //
+    CXPLAT_THREAD_CONFIG Config = { ThreadFlags, PartitionIndex, "DelayWorker", WorkerThread, this };
+    if (QUIC_FAILED(Thread.Create(&Config))) {
+        WriteOutput("CxPlatThreadCreate FAILED\n");
+        return false;
+    }
+    Initialized = true;
+    return true;
+}
+
+void DelayWorker::Shutdown()
+{
+    Shuttingdown = true;
+
+    if (Initialized) {
+        WakeWorkerThread();
+        Initialized = false;
+        //
+        // delete any pending work items
+        //
+        Lock.Acquire();
+        DelayedWorkContext* CurrentWorkItem = WorkItems;
+        while (nullptr != CurrentWorkItem) {
+            DelayedWorkContext* NextWorkItem;
+            if (WorkItemsTail == &CurrentWorkItem->Next) {
+                NextWorkItem = nullptr;
+            }
+            else {
+                NextWorkItem = CurrentWorkItem->Next;
+            }
+            delete CurrentWorkItem;
+            CurrentWorkItem = NextWorkItem;
+        }
+        WorkItems = nullptr;
+        Lock.Release();
+    }
+}
+
+void DelayWorker::WakeWorkerThread() {
+    if (!InterlockedFetchAndSetBoolean(&ExecutionContext.Ready)) {
+            CxPlatEventSet(WakeEvent);
+    }
+}
+
+CXPLAT_THREAD_CALLBACK(DelayWorker::WorkerThread, Context)
+{
+    DelayWorker* This = (DelayWorker*)Context;
+    CXPLAT_EXECUTION_STATE DummyState = {
+        0, 0, 0, UINT32_MAX, 0, CxPlatCurThreadID()
+    };
+    while (DelayedWork(This, &DummyState)) {
+        if (!InterlockedFetchAndClearBoolean(&This->ExecutionContext.Ready)) {
+            CxPlatEventWaitForever(This->WakeEvent); // Wait for more work
+        }
+    }
+    CXPLAT_THREAD_RETURN(0);
+}
+
+BOOLEAN
+DelayWorker::DelayedWork(
+    _Inout_ void* Context,
+    _Inout_ CXPLAT_EXECUTION_STATE* State
+)
+{
+    DelayWorker* This = (DelayWorker*)Context;
+    if (This->Shuttingdown) {
+        CxPlatEventSet(This->DoneEvent);
+        return FALSE;
+    }
+
+    DelayedWorkContext* WorkItem = nullptr;
+    This->Lock.Acquire();
+    if (This->WorkItems) {
+        WorkItem = This->WorkItems;
+        This->WorkItems = WorkItem->Next;
+        if (This->WorkItemsTail == &WorkItem->Next) {
+            This->WorkItemsTail = &This->WorkItems;
+        }
+        WorkItem->Next = nullptr;
+    }
+    This->Lock.Release();
+
+    if (nullptr != WorkItem) {
+        This->m_Server->SimulateDelay();
+        This->m_Server->SendResponse(WorkItem->Context, WorkItem->Handle, WorkItem->IsTcp);
+        delete WorkItem;
+        InterlockedFetchAndSetBoolean(&This->ExecutionContext.Ready); // We just did work, let's keep this thread hot.
+        State->NoWorkCount = 0;
+    }
+
+    return TRUE;
+}
+
+bool
+DelayWorker::QueueWork(
+    _In_ StreamContext* Context,
+    _In_ void* Handle,
+    _In_ bool IsTcp
+    )
+{
+    DelayedWorkContext* Work = new (std::nothrow) DelayedWorkContext();
+    if (nullptr == Work) {
+        return false;
+    }
```

Comment:
Since this is test/tool code, we _generally_ assume allocation success. Let's remove this check, which simplifies the whole function (it can't fail then) as well as the callers.